### PR TITLE
fix: Logout on API token no longer availlable remotely

### PR DESCRIPTION
**Pull request by adrien-coye**
### Abstract

A bug was found where an API token no longer available would not log out the user.
This is not a great experience, since all requests will fail.

This brings back code from `Mail` for a correct error handling of `infinite tokens` no longer availlable.

The navigation code was modified a bit to make sure the app cleanly logs out.

### CODE REVIEW

File: kDriveCore/Data/Cache/AccountManager.swift [hunk 1]
```diff
@@ -434,4 +440,21 @@ public class AccountManager: RefreshTokenDelegate, AccountManageable {
     public func account(for userId: Int) -> Account? {
         return accounts.first { $0.userId == userId }
     }
+
+    public func switchToNextAvailableAccountOrLogout() {
+        Task { @MainActor in
```
Comment (adrien-coye) [inline kDriveCore/Data/Cache/AccountManager.swift:445]: I need to run this on the main queue for navigation code to work well.
This is refactored on the APIV3 branch, so I won't refactor further here.

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation+Error.swift [hunk 2]
```diff
@@ -145,8 +147,21 @@ extension UploadOperation {
     @discardableResult
     func handleRemoteErrors(error: Error) -> Bool {
         guard let error = error as? DriveError, (error.type == .networkError) || (error.type == .serverError) else {
-            Log.uploadOperation("error:\(error) not a remote one ufid:\(uploadFileId)")
-            return false
+            // TODO: App level interceptor when the auth stack was unable to keep a valid session to log out
```
Comment (adrien-coye) [inline kDriveCore/Data/UploadQueue/Operation/UploadOperation+Error.swift:150]: Remove, this was a POC

**Review by adrien-coye — COMMENTED** [review]
self assesment

File: kDriveCore/Data/Cache/AccountManager.swift [hunk 3]
```diff
@@ -245,13 +248,16 @@ public class AccountManager: RefreshTokenDelegate, AccountManageable {
...
-            delegate?.currentAccountNeedsAuthentication()
-            notificationHelper.sendDisconnectedNotification()
-        }
+
+        DownloadQueue.instance.suspendAllOperations()
+
+        @InjectService var uploadQueue: UploadQueue
+        uploadQueue.suspendAllOperations()
+
+        delegate?.currentAccountNeedsAuthentication()
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDriveCore/Data/Cache/AccountManager.swift:258]: Done, only touch UI if the account matches.

File: kDrive/AppDelegate.swift [hunk 4]
```diff
@@ -463,9 +463,7 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDeleg
     // MARK: - Account manager delegate
 
     func currentAccountNeedsAuthentication() {
-        Task { @MainActor in
-            setRootViewController(SwitchUserViewController.instantiateInNavigationController())
-        }
+        Log.appDelegate("currentAccountNeedsAuthentication", level: .error)
```
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate.swift:466]: As this is no longer used, it should be removed.

Comment (adrien-coye) [reply kDrive/AppDelegate.swift]: done

File: kDriveCore/Data/Cache/AccountManager.swift [hunk 5]
```diff
@@ -434,4 +451,22 @@ public class AccountManager: RefreshTokenDelegate, AccountManageable {
     public func account(for userId: Int) -> Account? {
         return accounts.first { $0.userId == userId }
     }
+
+    public func logoutCurrentAccountAndSwitchToNextIfPossible() {
+        Task { @MainActor in
+            if let currentAccount {
+                removeTokenAndAccount(account: currentAccount)
+                notificationHelper.sendDisconnectedNotification()
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/Cache/AccountManager.swift:459]: This displays a snackbar or a notification when the user disconnect manually (from the menu) but this wasn't the case before

Comment (adrien-coye) [reply kDriveCore/Data/Cache/AccountManager.swift:459]: We displayed a snackbar when logged out automatically. I figured I could put it in the factorised method.
I got you, we only use the snackbar when the logout is automatic. Yeah we have enough snackbar already.

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**